programs/stake/src/stake_state.rs

```diff
@@ -1146,23 +1148,137 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
     }
 }
 
-fn get_info_if_mergable(
-    stake_keyed_account: &KeyedAccount,
-    clock: &Clock,
-    stake_history: &StakeHistory,
-) -> Result<Meta, InstructionError> {
-    match stake_keyed_account.state()? {
-        StakeState::Stake(meta, stake) => {
-            // stake must be fully de-activated
-            if stake.stake(clock.epoch, Some(stake_history), true) != 0 {
-                return Err(StakeError::MergeActivatedStake.into());
+#[derive(Clone, Debug, PartialEq)]
+enum MergeKind {
+    Inactive(Meta, u64),
+    ActivationEpoch(Meta, Stake),
+    FullyActive(Meta, Stake),
+}
+
+impl MergeKind {
+    fn meta(&self) -> &Meta {
+        match self {
+            Self::Inactive(meta, _) => meta,
+            Self::ActivationEpoch(meta, _) => meta,
+            Self::FullyActive(meta, _) => meta,
+        }
+    }
+
+    fn active_stake(&self) -> Option<&Stake> {
+        match self {
+            Self::Inactive(_, _) => None,
+            Self::ActivationEpoch(_, stake) => Some(stake),
+            Self::FullyActive(_, stake) => Some(stake),
+        }
+    }
+
+    fn get_if_mergeable(
+        stake_keyed_account: &KeyedAccount,
+        clock: &Clock,
+        stake_history: &StakeHistory,
+    ) -> Result<Self, InstructionError> {
+        match stake_keyed_account.state()? {
+            StakeState::Stake(meta, stake) => {
+                // stake must not be in a transient state. Transient here meaning
+                // activating or deactivating with non-zero effective stake.
+                match stake.delegation.stake_activating_and_deactivating(
+                    clock.epoch,
+                    Some(stake_history),
+                    true,
+                ) {
+                    /*
+                    (e, a, d): e - effective, a - activating, d - deactivating */
```

Comment:
Heading comment for tuple patterns in `get_if_mergeable()` match.

Wonky use of multi-line comment to trick `cargo fmt`